orttraining/orttraining/python/training/ort_triton/_sorted_graph.py

```diff
@@ -0,0 +1,210 @@
+        self._elementwise_graph_outputs: Set[str] = set()
+        for node in self._graph.node:
+            if is_elementwise_node(node):
+                self._elementwise_graph_outputs.update(node.output)
+
+        # Topological sort the nodes in the graph.
+        self._sorted_nodes: List[NodeProto] = topological_sort(
+            [input.name for input in self._graph.input] + [initializer.name for initializer in self._graph.initializer],
+            self._graph.node,
+        )
+
+        self._node_arg_infos: Dict[str, TensorInfo] = {}
+        for idx, input in enumerate(self._graph.input):
+            self._node_arg_infos[input.name] = TensorInfo(input.type.tensor_type.elem_type, self._input_shapes[idx])
+        for initializer in self._graph.initializer:
+            self._node_arg_infos[initializer.name] = TensorInfo(
+                initializer.data_type,
+                list(to_numpy_array(initializer).shape),
+            )
+
+        # Decompose complex operators.
+        self._decompose()
+
+        # Sort the initializers in reference order.
+        # We try to reuse Triton module for different ONNX models with same graph structure,
+        # even the node args names in the models are different.
+        # Sorting the initializers can help to generate same model key for different ONNX models.
+        initializers = {}
+        for initializer in self._graph.initializer:
+            initializers[initializer.name] = initializer
+        self._sorted_initializers: List[TensorInfo] = []
+        for node in self._sorted_nodes:
+            for input in node.input:
+                if input in initializers:
+                    self._sorted_initializers.append(initializers[input])
+                    initializers.pop(input)
+
+        # Split nodes to constant nodes and non-constant nodes.
+        self._const_nodes: List[NodeProto] = [node for node in self._sorted_nodes if node.op_type == "Constant"]
+        self._sorted_nodes: List[NodeProto] = [node for node in self._sorted_nodes if node.op_type != "Constant"]
+
+    def __str__(self):
+        """
+        Generate a unique key for the model based on the graph structure, ignoring the node args names.
+        We try to reuse Triton module for different ONNX models with same graph structure.
+        """
+        graph_inputs = []
+        name_map = {}
+        for idx, input in enumerate(self._graph.input):
+            shape_str = str(self._input_shapes[idx]).replace(" ", "")
+            graph_inputs.append(f"({str(input.type.tensor_type.elem_type)},{shape_str})")
+            name_map[input.name] = f"i{idx}"
+        graph_inputs_str = ",".join(graph_inputs)
+
+        constants = []
+        for idx, initializer in enumerate(self._sorted_initializers):
+            data_str = np.array2string(to_numpy_array(initializer), separator=",").replace("\n", "").replace(" ", "")
```

Comment:
Why not using only the shape and the type here?

Reply:
Generated code will put initializer's value (currently supports only scalar) as constexpr in the code, different value will have different generated code.